interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,151 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - gadget
+    allow-connection:
+      plug-attributes:
+        shared-memory: $SLOT(shared-memory)
+    deny-auto-connection: true
+`
+
+// sharedMemoryInterface allows sharing sharedMemory between snaps
+type sharedMemoryInterface struct{}
+
+func (iface *sharedMemoryInterface) Name() string {
+	return "shared-memory"
+}
+
+func (iface *sharedMemoryInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:              sharedMemorySummary,
+		BaseDeclarationSlots: sharedMemoryBaseDeclarationSlots,
+		AffectsPlugOnRefresh: true,
+	}
+}
+
+func (iface *sharedMemoryInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	sharedMemory, ok := slot.Attrs["shared-memory"].(string)
+	if !ok || sharedMemory == "" {
```

Comment:
Shouldn't we throw an error if the attribute is not a string?

Reply:
Yes, that would be more correct, on the other hand I copied the code from https://github.com/snapcore/snapd/blob/master/interfaces/builtin/content.go#L74-L81 so I thought this would be fine too :-)

But I'll add the error return.

Reply:
This seems to work fine. Let me know if you'd like me to make this change, or if you can suggest some improvement:
```go
    sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]                                             
    sharedMemory, ok := sharedMemoryAttr.(string)
    if isSet && !ok {                                                                                  
        return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,          
            slot.Attrs["shared-memory"])
    }       
    if sharedMemory == "" {                                                                            
```

Reply:
Looks reasonable, one idea to consider but no strong opinion if it's better:
```   
    var sharedMemory string
    sharedMemoryAttr, isSet := slot.Attrs["shared-memory"]                                             
    if isSet {
        var ok bool
        sharedMemory, ok = sharedMemoryAttr.(string) 
        if !ok {                                                                           
            return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,          
               slot.Attrs["shared-memory"])
        }
    }       
    if sharedMemory == "" { 
```                                                                       

We do want to provide a default value right? Because otherwise we could use .Attr(..) getter which does type-checks - however it also complains about missing attribute (but doesn't have distinct error types).

Reply:
Yes, the whole point is to gracefully allow for the attribute to be missing. I'll go for my version of the code then, as I find it easier to read.

Reply:
Done